### ISSUE TITLE
Enable user keybindings to forcefully override plugin keybindings

### ISSUE THREAD

**Issue by nicolasdumitru** [neovim/neovim#25101]
### Problem

Some plugins have very invasive methods of setting keybindings. This can cause a lot of frustration and confusion. Neovim should provide the tools needed to deal with this invasive behavior as elegantly as possible.

I had issues with is [Neorg](https://github.com/nvim-neorg/neorg), but any plugin can do this, so this is NOT a critique of Neorg (which I think is a great plugin otherwise). In order illustrate what I mean by invasive, I will describe what Neorg does:
In simple terms, by default, Neorg not only sets BUFFER-LOCAL keybindings, but it also does so REPETITIVELY when it's in use. This has the following (frustrating) effects:

1. It overrides normal (not necessarily 'normal' mode) user-defined keybindings if they coincide with Neorg-specific keybindings;
2. Neorg keybindings cannot be overriden in the `after/` directory (the usual way);
3. Disabling one of these keybindings with `:unmap <buffer> <some-Neorg-keybinding>` only works if you don't press that keybinding again - rendering the `:unmap` command COMPLETELY useless.

This is made even worse by the fact that, in order to avoid this kind of plugin behavior:

- **generally** - the plugin has to give the user sane options for disabling it;
- **in this case** - the user has to either disable all of the keybindings OR use plugin-specific syntax to override the plugins. This means you are either familiar with all of them and disable what you don't like or wait until something breaks while you're trying to do work.

You can read more about this in [this issue](https://github.com/nvim-neorg/neorg/issues/1023).

I would like to reiterate that this is a Neovim problem, **NOT** a plugin problem (although plugins shouldn't do this) and I am not here to denigrate Neorg.

### Expected behavior

The most elegant solution that I can think of is adding a new special map-argument - "force" (or whatever suggestive name you think is more appropriate).

The "force" keybinding argument should:

- only be usable by the user in command mode and in the user configuration directories (`$XDG_CONFIG_HOME/nvim`, `$XDG_CONFIG_DIRS[1]/nvim`, `$XDG_CONFIG_DIRS[2]/nvim`, `$XDG_CONFIG_DIRS[2]/nvim/after`, `$XDG_CONFIG_DIRS[1]/nvim/after`, `$XDG_CONFIG_HOME/nvim/after`
, as `:h rtp` lists them), **NOT** all the directories in the 'runtimepath'. If plugins can use it, it defeats its own purpose.
- determine Neovim to **only** do what the user defined (as specified above) when the keybinding is pressed. Not anything else.
- not cause errors. Plugins may attempt to set whatever keybindings they have to, but any one of those should simply not work when the user defines the same keybinding with the "force" argument.

As an example, setting:
`vim.keymap.set("n", "<leader>m", <user_command>, { force = true }) -- notice it's a global keybinding`
in a user's `$XDG_CONFIG_HOME/nvim/init.lua` and loading a plugin that would normally, while in use, repetitively set:
`vim.keymap.set("n", "<leader>m", <plugin_command>, { buffer = true }) -- notice it's a buffer-local keybinding` means that `<leader>m` never executes `plugin_command` and always executes `user_command` when pressed, no matter how many times the plugin tries to set it and irrespective of the fact that the plugin's keybinding is buffer-local.

In layman's terms: if you set a keybinding yourself with `{ force = true }`, you can be sure that Neovim is going to only do what you set (manually or in your config) and nothing else.

I think that the new map-argument would be an elegant solution to the problem and would give users more control over their plugins and their keybindings. This is very relevant because it is in line with Neovim's goals of being extensible and usable. "Just don't use the plugin" is not a solution to the problem - Neovim is supposed to be extensible (play nice with plugins and provide the tools needed to make plugins play nice - within reason) and some plugins are unique and there are no similar replacements.

**Comment by justinmk** [comment]
> The "force" keybinding argument should:  only be usable by the user in command mode and in the user configuration directories

Filing this under [plugin namespacing](https://github.com/neovim/neovim/issues/16313).

> I think that the new map-argument would be an elegant solution to the problem and would give users more control over their plugins and their keybindings.

On the contrary, it looks like an XY problem to me, or tunnel vision. Plugins can do all kinds of stuff.

And if this is implemented, the next request will be "let plugins the "force" argument... 

> Neovim's goals of being extensible and usable.

Trotting that out as a catch-all justification is... ahem. And it's contradictory here since you are asking for the extensibility to be limited. Usability is also a central goal, but that doesn't justify every random feature request since it needs to be weighed against other work.

**Comment by nicolasdumitru** [comment]
@justinmk 
> it looks like an XY problem to me, or tunnel vision

I've thought about this and even hesitated before opening a feature request. I don't agree that it's an XY problem and I think I have provided enough arguments as to why it's a Neovim problem, not a plugin problem. Furthermore, it seems to me like the XY problem seems to imply that there is already a better way to solve the problem. If it's not the "don't use a pluginle" kind of solution, please state it and I will happily implement it.

> it looks like an XY problem to me, or tunnel vision

How is giving users more control over keybindings an XY problem?

> And it absolutely is critical that users insist on well-behaved plugins, not private workarounds.

'Well-behaved' is vague at best. The fact that this situation even exists proves irrefutably that your definition of well-behaved is not universally accepted. You can't expect every plugin to do exactly what you want by default. You don't expect your editor to do so either. This is why we configure them. How is it a workaround to give users the option of making sure what they set in their configs is exactly what Neovim does?

> Trotting that out as a catch-all justification is... ahem

It's not a catch-all. Part of extensibility is choosing what you want to extend. Otherwise, you'd just use some of the "full featured" programs that do everything under the sun poorly and call it a day. What point is there in "extending" your editor in a way that you don't want it to be extended?

> And it's contradictory here since you are asking for the extensibility to be limited.

I'm not asking for the extensibility to be limited. I have never asked for such a thing. This is an intentional misinterpretation of what I said. I have asked for the ability to set boundaries. This is not contradicting anything.

You added the "closed:duplicate" label. Please indicate what issue this feature request is a duplicate of. Is it #16313?

Unfortunately, this written form of the message may not convey the tone of voice that I wish it does. To make this abundantly clear: I am not trying to argue - I have no interest in that; I am trying to help. If you think I'm wrong, please tell me in what way I'm wrong.

**Comment by lewis6991** [comment]
I totally agree with Justin on this.

> I don't agree that it's an XY problem and I think I have provided enough arguments as to why it's a Neovim problem, not a plugin problem. 

Sorry, but I don't think the arguments have much strength and to me appear quite narrowly focused.

Neorg is not being _well-behaved_ in this case. If the user sets up mappings, then Neorg shouldn't clobber them. If it insists on it, then that's just a limitation of Neorg you need to live with and doesn't justify adding some convoluted mechanism to avoid this.

> I have asked for the ability to set boundaries. 

This doesn't align with anything in Neovim. Neovim provides extensibility, not guard railing. If a plugin does something you don't want it to, either avoid using the plugin or change it.

> The "force" keybinding argument should:
>
> -    only be usable by the user in command mode and in the user configuration directories ($XDG_CONFIG_HOME/nvim, $XDG_CONFIG_DIRS[1]/nvim, $XDG_CONFIG_DIRS[2]/nvim, $XDG_CONFIG_DIRS[2]/nvim/after, $XDG_CONFIG_DIRS[1]/nvim/after, $XDG_CONFIG_HOME/nvim/after
    , as :h rtp lists them), NOT all the directories in the 'runtimepath'. If plugins can use it, it defeats its own purpose.

Sorry but this is not a good proposal. Neovim is completely agnostic to anything in `runtimepath`. This also means there is zero distinction between user config and plugins. Adding any form of special casing is messy, confusing and unnecessary.

> -   determine Neovim to only do what the user defined (as specified above) when the keybinding is pressed. Not anything else.

What does this even mean? This seems to imply that you think user config and plugins are distinct, which is not the case.

> -   not cause errors. Plugins may attempt to set whatever keybindings they have to, but any one of those should simply not work when the user defines the same keybinding with the "force" argument.

huh? Keybindings with `force` should silent any errors now? This really makes no sense at all. Errors are a runtime problem and can happen at any stage in a script, either at the start or just before the end. Are you expecting Neovim to backtrack on actions a keymapping has performed up to the error?

**Comment by nicolasdumitru** [comment]
> This doesn't align with anything in Neovim. Neovim provides extensibility, not guard railing. If a plugin does something you don't want it to, either avoid using the plugin or change it.

This is not as easy as it sounds.

@lewis6991 I was, indeed, under the impression that user config and plugins are distinct. I now understand that what I thought should have been fairly trivial in terms of implementation is actually fairly complex. This is not what I had hoped for, but I'll deal with it. Thanks for the clarification.

**Comment by lewis6991** [comment]
What might be feasible is copying the `default` feature from `:highlight`.

```vimhelp
					*:hi-default* *:highlight-default*
The [default] argument is used for setting the default highlighting for a
group.	If highlighting has already been specified for the group the command
will be ignored.  Also when there is an existing link.

Using [default] is especially useful to overrule the highlighting of a
specific syntax file.  For example, the C syntax file contains:  
	:highlight default link cComment Comment
If you like Question highlighting for C comments, put this in your vimrc file:  
	:highlight link cComment Question
Without the "default" in the C syntax file, the highlighting would be
overruled when the syntax file is loaded.
```

So plugins could _choose_ to apply their mappings with `default = true` as to not clobber any user mappings. This would have the same effect as:

```lua
if vim.fn.hasmapto(...) == 0 then
    vim.keymap.set(...)
end
```

but just more convenient with less boilerplate.

**Comment by justinmk** [comment]
> I was, indeed, under the impression that user config and plugins are distinct. I now understand that what I thought should have been fairly trivial in terms of implementation is actually fairly complex.

Yeah that's why https://github.com/neovim/neovim/issues/16313#issuecomment-968265007 covers this topic. The general idea is that plugins (and user config) can "own" a namespace. Just like an autocmd group can be deleted with one command, you could delete a plugin's mappings with one command.
